dot/core/sync.go

```diff
@@ -0,0 +1,143 @@
+package core
+
+import (
+	"encoding/binary"
+	"errors"
+	"math/big"
+	mrand "math/rand"
+	"sync"
+	"time"
+
+	"golang.org/x/exp/rand"
+
+	"github.com/ChainSafe/gossamer/dot/network"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/common/optional"
+
+	log "github.com/ChainSafe/log15"
+)
+
+// Syncer deals with chain syncing by sending block request messages and watching for responses.
+type Syncer struct {
+	blockState    BlockState             // retrieve our current head of chain from BlockState
+	blockNumberIn <-chan *big.Int        // incoming block numbers seen from other nodes that are higher than ours
+	msgOut        chan<- network.Message // channel to send message to network service
+	lock          *sync.Mutex
+	synced        bool
+}
+
+// SyncerConfig is the configuration for the Syncer.
+type SyncerConfig struct {
+	BlockState    BlockState
+	BlockNumberIn <-chan *big.Int
+	MsgOut        chan<- network.Message
+	Lock          *sync.Mutex
+}
+
+// NewSyncer returns a new Syncer
+func NewSyncer(cfg *SyncerConfig) (*Syncer, error) {
+	if cfg.BlockState == nil {
+		return nil, errors.New("cannot have nil BlockState")
+	}
+
+	if cfg.BlockNumberIn == nil {
+		return nil, errors.New("cannot have nil BlockNumberIn channel")
+	}
+
+	if cfg.MsgOut == nil {
+		return nil, errors.New("cannot have nil MsgOut channel")
+	}
+
+	return &Syncer{
+		blockState:    cfg.BlockState,
+		blockNumberIn: cfg.BlockNumberIn,
+		msgOut:        cfg.MsgOut,
+		lock:          cfg.Lock,
+		synced:        true,
+	}, nil
+}
+
+// Start begins the syncer
+func (s *Syncer) Start() {
+	go s.watchForBlocks()
+}
+
+func (s *Syncer) watchForBlocks() {
+	for {
+		peerNum := <-s.blockNumberIn
```

Comment:
`peerNum` ?

Reply:
it's the peer's block num, will update it to something better